### PR TITLE
[assessment-tool] Added view for non-found results.

### PR DESCRIPTION
**Pull request by ygorshenin**
Added a simple list view for results that are present in the search samples json, but weren't retrieved by the search engine. Relevances of these results may be modified by the user and these modifications will be saved.

@milchakov @bykoianko PTAL

### CODE REVIEW

File: search/search_quality/assessment_tool/main_view.cpp [hunk 1]
```diff
@@ -242,8 +278,15 @@ void MainView::InitDocks()
           [this]() { m_model->OnShowPositionClicked(); });
 
   {
-    auto * model = m_sampleView->GetResultsView().selectionModel();
-    connect(model, &QItemSelectionModel::selectionChanged, this, &MainView::OnResultSelected);
+    auto const & view = m_sampleView->GetFoundResultsView();
+    connect(&view, &ResultsView::OnResultSelected,
+            [&](int index) { m_model->OnResultSelected(index); });
```
Comment (milchakov) [inline search/search_quality/assessment_tool/main_view.cpp]: а почему захватываешь всё, а не this?

File: search/search_quality/assessment_tool/result_view.cpp [hunk 2]
```diff
@@ -33,16 +35,34 @@ void SetText(QLabel & label, std::string const & text) {
...
 }
 
+ResultView::ResultView(search::Result const & result, QWidget & parent)
+  : ResultView(result.GetString(), result.GetFeatureType(), result.GetAddress(), parent)
+{
+}
+
+ResultView::ResultView(search::Sample::Result const & result, QWidget & parent)
+  : ResultView(strings::ToUtf8(result.m_name), GetResultType(result), string() /* address */,
+               parent)
```
Comment (milchakov) [inline search/search_quality/assessment_tool/result_view.cpp]: используй пожалуйста вместо `string()` `{}` или `""`, тогда перенос строки не потребуется

Comment (ygorshenin) [inline search/search_quality/assessment_tool/result_view.cpp]: Неее, я лучше буду использовать string().

Comment (ygorshenin) [inline search/search_quality/assessment_tool/result_view.cpp]: Переносы строк меня не пугают.

File: search/search_quality/assessment_tool/main_view.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/search_quality/assessment_tool/main_view.cpp]: Давай this.

**Comment by ygorshenin** [comment]
@bykoianko @milchakov PTAL

File: search/search_quality/assessment_tool/context.cpp [hunk 3]
```diff
@@ -130,3 +153,12 @@ void ContextList::ApplyEdits()
...
     context.ApplyEdits();
 }
+
+void ContextList::OnContextUpdated(size_t index)
+{
+  if (!m_hasChanges[index] && m_contexts[index].HasChanges())
+    ++m_numChanges;
+  if (m_hasChanges[index] && !m_contexts[index].HasChanges())
+    --m_numChanges;
+  m_hasChanges[index] = m_contexts[index].HasChanges();
```
Comment (bykoianko) [inline search/search_quality/assessment_tool/context.cpp]: CHECK_LESS_OR_EQUAL(0, m_numChanges, ())?

File: search/search_quality/assessment_tool/main_view.cpp [hunk 4]
```diff
@@ -58,38 +60,63 @@ void MainView::SetSamples(ContextList::SamplesSlice const & samples)
...
-  m_sampleView->ShowResults(begin, end);
+  m_sampleView->ShowNonFoundResults(results);
 }
 
 void MainView::MoveViewportToResult(search::Result const & result)
 {
   m_framework.ShowSearchResult(result, false /* animation */);
 }
 
+void MainView::MoveViewportToResult(search::Sample::Result const & result) {
```
Comment (bykoianko) [inline search/search_quality/assessment_tool/main_view.cpp]: } from new line?

File: search/search_quality/assessment_tool/main_view.cpp [hunk 5]
```diff
@@ -58,38 +60,63 @@ void MainView::SetSamples(ContextList::SamplesSlice const & samples)
...
+  int constexpr kViewportAroundResultSizeM = 100;
```
Comment (bykoianko) [inline search/search_quality/assessment_tool/main_view.cpp]: double?

File: search/search_quality/assessment_tool/context.cpp (hunk 3 above)
Comment (ygorshenin) [inline search/search_quality/assessment_tool/context.cpp]: m_numChanges is a size_t. So, this CHECK is senseless.

File: search/search_quality/assessment_tool/main_view.cpp (hunk 4 above)
Comment (ygorshenin) [inline search/search_quality/assessment_tool/main_view.cpp]: Done

File: search/search_quality/assessment_tool/main_view.cpp (hunk 5 above)
Comment (ygorshenin) [inline search/search_quality/assessment_tool/main_view.cpp]: Why? What's wrong with int?

**Comment by ygorshenin** [comment]
@bykoianko PTAL

File: search/search_quality/assessment_tool/context.cpp (hunk 3 above)
Comment (bykoianko) [inline search/search_quality/assessment_tool/context.cpp]: Ok

File: search/search_quality/assessment_tool/main_view.cpp (hunk 5 above)
Comment (bykoianko) [inline search/search_quality/assessment_tool/main_view.cpp]: Nothing wrong. Just MercatorBounds::RectByCenterXYAndSizeInMeters() accepts double. It's up to you.

**Pull request merged**